File: src/121-best-time-to-buy-and-sell-stock/solution.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
  min, max = 0, 1 
  n = prices.length
  res = 0

  while max < n
    sum = prices[max] - prices[min]
    res = [res, sum].max
    min = max if prices[max] < prices[min]
    max += 1
  end

  res
end

File: src/121-best-time-to-buy-and-sell-stock/solution.ts
function maxProfit(prices: number[]): number {
  let l: number = 0, r: number = 1, res: number = 0, sum: number = 0;
  while(r < prices.length) {
    sum = prices[r] - prices[l];
    res = Math.max(sum, res);
    if(prices[r] < prices[l]) l = r;
    r++;
  }

  return res;
};
